teps/0089-nonfalsifiable-provenance-support.md

```diff
@@ -0,0 +1,255 @@
+
+The solution to this is Signed TaskRuns, where the TaskRun has to be signed and verified every time it has been modified.
+That way, we can prove that the TaskRun hasn't been tampered with during or after execution.
+
+The Tekton Controller will need to request a signature from SPIRE each time the TaskRun is modified to prove that it hasn't been tampered with during execution.
+Roughly, this will look something like this:
+* Controller initiates a TaskRun, and stores a signature over its contents as an annotation
+
+Each time the TaskRun is modified,
+* Controller verifies the TaskRun hasn't been modified
+* Controller requests a new SPIRE SVID and signature over the new modified TaskRun
+* Controller stores the new SVID and signature on the TaskRun as an annotation
+
+Right now, the design details around this are a little fuzzy!
+The plan is to first implement signed Results, and then if this design no longer seems like the correct fit for signed TaskRuns then it will be revisited in this TEP :)
+
+Things to keep in mind when designing this feature further:
+* There could potentially be a performance implication here, if we need to request signatures every time the TaskRun is modified during execution.
+* We may want to ignore labels/annotations when checking for modifications, since they might be updated by other services
+* What if mutating admission controllers are intentionally changing TaskRuns (e.g. Solarwinds injecting Tasks into a Pipeline)
+
+### 2. Tekton Pipelines can't verify that the results it reads weren't modified
+The solution to this is Signed Results.
+We will modifiy the entrypointer image to sign results with SPIRE once they're available.
+The signature and SVID provided by SPIRE will be stored on the TaskRun itself as annotations for a client to verify (this could be Pipelines or Chains).
+
+
+## Motivation
+
+Security! =)
+
+We'll also need this feature for Tekton to achieve [SLSA 3](https://slsa.dev/levels), which requires Non-falsifiable Provenance.
+
+### Requirements
+* SPIRE is configured to only issue SVIDs to the Tekton controller via a [k8s Workload Attestor](https://github.com/spiffe/spire/blob/main/doc/plugin_agent_workloadattestor_k8s.md)
+
+### Goals
+
+* Results can be verified as non-falsifiabe
+* A TaskRun yaml can be verified as non-falsifiable
+* Clients are confident that TaskRuns were initiated and monitored by the Tekton controller
+
+### Non-Goals
+* Protect against a malicious cluster admin (this will be a goal for SLSA 4)
+
+
+
+## Proposal
+As mentioned above, the basic design looks like this (this is meant to be high level and still needs to be fleshed out a bit):
+1. Tekton Pipelines receives a TaskRun config, and generates the Pod for it with SPIRE mounted in
+1. The Pod executes, and the entrypointer requests an SVID and signature over the Results
+1. Tekton Pipelines verifies the Results
+1. Meanwhile, Tekton Pipelines has been verifying that the TaskRun hasn't been modified during execution
```

Comment:
please elaborate a little on `TaskRun`, I think `TaskRun` here means task specifications (including input params) either through `taskRef` or `taskSpec` 🤔  Or does it also include `status` since `status` is modified by the pipeline controller until the execution is over.

Reply:
i basically meant it to include the entire TaskRun yaml that's being executed, and i'm guessing the `status` should be included as well.